### ISSUE TITLE
[no-implied-eval] Function probably treated as string if refer via 'this'

### ISSUE THREAD

**Issue by CyberCookie** [typescript-eslint/typescript-eslint#3391]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {}
}
```

```TS
const someObj = { // eslint-disable-line
    someFunc() {}, // eslint-disable-line

    withTimeout() {
        setTimeout(this.someFunc)
    }
}
```

**Expected Result**
No errors

**Actual Result**
Error: Implied eval. Consider passing a function.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.23.0` |
| `@typescript-eslint/parser`        | `4.23.0` |
| `TypeScript`                       | `4.2.4` |
| `ESLint`                           | `7.26.0` |
| `node`                             | `14.16.1` |

**Comment by bradzacher** [comment]
I'm unable to reproduce this against master
![image](https://user-images.githubusercontent.com/7462525/118372314-e4ce9400-b565-11eb-9d65-931378bb3226.png)

Could you please create an isolated repro repo?

**Comment by CyberCookie** [comment]
@bradzacher  Thank you for the quick reply!

Sorry, in the example above I forgot to include the rule causing the error.
Could you please check again with:

```
{
  "rules": {
    "@typescript-eslint/no-implied-eval": "error"
  }
}
```
I will create repro repo if the mentioned error will not appear for the second attempt...

![image](https://user-images.githubusercontent.com/9907516/118388383-d26b5d80-b62c-11eb-9dc3-08edda41c32c.png)

**Comment by bradzacher** [comment]
> Sorry, in the example above I forgot to include the rule causing the error.

You listed the rule in the issue title, so I tested the code directly against master against the tests for `no-implicit-eval`.

**Comment by CyberCookie** [comment]
@bradzacher I've created a repro repo: https://github.com/CyberCookie/typescript_eslint_issue_3391

**Comment by bradzacher** [comment]
if I were to hazard a guess - it's because your `tsconfig.eslint.json` has no compiler options turned on. Meaning it's got `strict: false` by default. Meaning that the `this` checks provided by `noImplicitThis` are not turned on.

But your `tsconfig.json` has `strict: true`.

Hence VSCode shows everything fine, but ESLint does not.

You should keep your `tsconfig.eslint.json` in sync with your `tsconfig.json` by using `extends`.

**Comment by CyberCookie** [comment]
i've managed to solve this by extending from `tsconfig.json` and added `noImplicitThis: true` !
Thanks!

Closing the issue
